python_modules/dagster/dagster/core/definitions/asset_layer.py

```diff
@@ -397,6 +398,15 @@ class AssetLayer:
             keys for each asset key produced by this job.
     """
 
+    _asset_keys_by_node_input_handle: Mapping[NodeInputHandle, AssetKey]
+    _asset_info_by_node_output_handle: Mapping[NodeOutputHandle, AssetOutputInfo]
+    _asset_deps: Mapping[AssetKey, AbstractSet[AssetKey]]
+    _dependency_node_handles_by_asset_key: Mapping[AssetKey, Set[NodeHandle]]
+    _source_assets_by_key: Mapping[AssetKey, "SourceAsset"]
+    _asset_defs_by_key: Mapping[AssetKey, "AssetsDefinition"]
+    _asset_defs_by_node_handle: Mapping[NodeHandle, Set["AssetsDefinition"]]
+    _io_manager_keys_by_asset_key: Mapping[AssetKey, str]
```

Comment:
A few reasons:

- This is the standard way to annotate instance attrs in Python-- see [PEP 526](https://peps.python.org/pep-0526/#class-and-instance-variable-annotations). There is a special `ClassVar` annotation you use for class variables.
- IMO it is a lot more readable than annotations in `__init__.

> I'm surprised the types didnt propagate from __init__ correctly though, i guess issues witch check calls?

- Yes, the types are often not inferred correctly in `__init__`. I _think_ it would work the same as the above style if we explicitly annotated the type for every assignment in `__init__`, but I'm not 100% sure. Regardless, it's a lot harder to read/find the declared type when its embedded in often lengthy `__init__` logic.
- The above style matches the style used by dataclasses and pydantic, which I still hope to transition to